Dockerfile

```diff
@@ -18,117 +20,129 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
+RUN apt-get update \
+  && apt-get install -y \
+    ruby \
+    ruby-dev \
+  && rm -rf /var/lib/apt/lists/*
+
+# Install Go.
+# This installs Go 1.6 on Ubuntu 16.04.
+RUN apt-get update \
+  && apt-get install -y golang \
+  && rm -rf /var/lib/apt/lists/*
 
+# Download the go protobuf support.
 ENV GOPATH /go
 ENV PATH $GOPATH/bin:/home/linuxbrew/.linuxbrew/opt/go/libexec/bin:$PATH
-RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
-RUN go get -u github.com/golang/protobuf/proto github.com/golang/protobuf/protoc-gen-go
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" \
+  && chmod -R 777 "$GOPATH" \
+  && go get -u github.com/golang/protobuf/proto github.com/golang/protobuf/protoc-gen-go
 
 # Install packman
 # TODO: consider installing released packages once artman is versioned (so that
 #   each release of artman is pegged to a release of packman)
 RUN npm install -g https://github.com/googleapis/packman.git
 
 # Setup tools for codegen of Ruby
-RUN gem install rubocop --version '= 0.39.0' --no-ri --no-rdoc
-RUN gem install bundler --version '= 1.12.1' --no-ri --no-rdoc
-RUN gem install rake --version '= 10.5.0' --no-ri --no-rdoc
-RUN gem install grpc-tools --version '=1.0.0' --no-ri --no-rdoc
+RUN gem install rake --no-ri --no-rdoc \
+  && gem install rubocop --version '= 0.39.0' --no-ri --no-rdoc \
+  && gem install bundler --version '= 1.12.1' --no-ri --no-rdoc \
+  && gem install rake --version '= 10.5.0' --no-ri --no-rdoc \
+  && gem install grpc-tools --version '=1.0.0' --no-ri --no-rdoc
 
 # Install couple of git repos
-WORKDIR /
 RUN git clone https://github.com/googleapis/googleapis
 RUN git clone https://github.com/googleapis/toolkit
 ENV TOOLKIT_HOME /toolkit
-WORKDIR /toolkit
-RUN sudo ./gradlew install # Install toolkit. Must sudo to download gradle plugins.
+
+# Install toolkit. Must sudo to download gradle plugins.
+RUN cd /toolkit \
+  && ./gradlew install \
+  && cd /
 
 # Install googleapis protocol compiler plugin which is needed to build gapic resource classes.
-RUN pip install -e git+https://github.com/googleapis/proto-compiler-plugin#egg=remote
-
-# Install PHP protobuf plugin
-WORKDIR /
-RUN apt-get install -y ruby2.0 ruby2.0-dev
-RUN /usr/bin/gem2.0 install ronn --no-ri --no-rdoc
-RUN git clone https://github.com/stanley-cheung/Protobuf-PHP.git
-WORKDIR /Protobuf-PHP
-RUN rake pear:package version=1.0 --trace
-RUN pear install Protobuf-1.0.tgz
-
-# Install PHP formatting tools
-RUN pear install PHP_CodeSniffer-2.7.0
-RUN curl -L https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases/download/v1.11.8/php-cs-fixer.phar -o /usr/local/bin/php-cs-fixer
-RUN chmod a+x /usr/local/bin/php-cs-fixer
-
-# Install artman.
-# TODO(ethanbao): pipeline should be installed from package manager.
-ADD . /src
-WORKDIR /src
-RUN pip install -e .
+RUN pip3 install -e git+https://github.com/googleapis/proto-compiler-plugin#egg=remote
+
+# Install PHP protobuf plugin.
+RUN gem install ronn --no-ri --no-rdoc \
```

Comment:
gem does not (to the best of my knowledge) maintain a running list of all available packages in the same way that apt does.

When you run `apt-get update`, it literally pulls down metadata for _every single available package_ on those apt repos, and then uses that when you run `apt-cache search` or `apt-get install`.

I suspect the reason for this behavior is that this is where the multiple sources of truth (the N apt repositories you have added to `sources.list`) is resolved into a single list of packages; it would be too cumbersome to check each separate repo for a package every time you ran an install (especially given the high dependency count of apt packages).

By contrast, `gem install` (and pip and npm) check a single source of truth, and so no complex local cache is required.